### PR TITLE
fix: Allows checks for both underscores and dashes as equal

### PR DESCRIPTION
**Pull request by jshwi**
Fixes open issue: sync-then-check fails with underscores #72

### CODE REVIEW

File: tests/sync_pipfile_test.py [hunk 1]
```diff
@@ -54,3 +54,27 @@ def test_sync_dev_pipfile_no_original(tmp_path):
...
+)
+def test_sync_underscore_or_dash(
+    capsys, tmp_path, shared_datadir, source_pipfile_dirname, update_count
+):
+    """
+    sync --pipfile should work for either dash or underscore names.
+
+    Asserts fix for https://github.com/Madoshakalaka/pipenv-setup/issues/72.
+    """
+    pipfile_dir = shared_datadir / source_pipfile_dirname
```
Comment (bryant-finney) [inline tests/sync_pipfile_test.py:71]: This suggestion is paired with a couple additional ones

```suggestion
    pipfile_dir = shared_datadir / "dash_or_underscore_0"
```

File: tests/sync_pipfile_test.py [hunk 2]
```diff
@@ -54,3 +54,27 @@ def test_sync_dev_pipfile_no_original(tmp_path):
         text = setup_file.read_text()
         assert "pytest~=5.1" in text, text
         assert "requirementslib~=1.5" in text, text
+
+
+@pytest.mark.parametrize(
+    ("source_pipfile_dirname", "update_count"),
+    [("dash_or_underscore_0", 23)],
+)
```
Comment (bryant-finney) [inline tests/sync_pipfile_test.py:62]: When combined with [this suggestion](https://github.com/Madoshakalaka/pipenv-setup/pull/86/files#r752446810):

```suggestion
```

- the `update_count` parameter is not used by this test
- because this test only uses a single dataset, parametrizing the `source_pipfile_dirname` isn't necessary

File: tests/sync_pipfile_test.py [hunk 3]
```diff
@@ -54,3 +54,27 @@ def test_sync_dev_pipfile_no_original(tmp_path):
...
+def test_sync_underscore_or_dash(
+    capsys, tmp_path, shared_datadir, source_pipfile_dirname, update_count
+):
```
Comment (bryant-finney) [inline tests/sync_pipfile_test.py:65]: After [this suggestion](https://github.com/Madoshakalaka/pipenv-setup/pull/86/files#r752446810) and the one above:

```suggestion
def test_sync_underscore_or_dash(shared_datadir):  # type: pathlib.Path
```

File: tests/sync_pipfile_test.py [hunk 4]
```diff
@@ -54,3 +54,27 @@ def test_sync_dev_pipfile_no_original(tmp_path):
...
+    """
+    sync --pipfile should work for either dash or underscore names.
+
+    Asserts fix for https://github.com/Madoshakalaka/pipenv-setup/issues/72.
+    """
+    pipfile_dir = shared_datadir / source_pipfile_dirname
+    for filename in ("Pipfile", "Pipfile.lock", "setup.py"):
+        copy_file(pipfile_dir / filename, tmp_path)
+
```
Comment (bryant-finney) [inline tests/sync_pipfile_test.py:74]: This is unnecessary, as the [tests.conftest.data fixture](https://github.com/Madoshakalaka/pipenv-setup/blob/05d07f9c91ffca6b1364bef01cfce2a88097cc00/tests/conftest.py#L23) already performs this operation

```suggestion
```

File: tests/sync_pipfile_test.py [hunk 5]
```diff
@@ -54,3 +54,27 @@ def test_sync_dev_pipfile_no_original(tmp_path):
...
+    with data(str(pipfile_dir), tmp_path):
+        cmd(["", "sync", "--pipfile"])
+        cmd(["", "check"])
+
+    captured = capsys.readouterr()
+    assert "in pipfile but not in install_requires" not in captured.out
```
Comment (bryant-finney) [inline tests/sync_pipfile_test.py:80]: I don't believe this will ever be hit. If the `check` command fails, the `fatal_error()` function will either raise an exception or execute `sys.exit()`.

```suggestion
```

File: tests/sync_pipfile_test.py [hunk 6]
```diff
@@ -54,3 +54,27 @@ def test_sync_dev_pipfile_no_original(tmp_path):
...
+    """
+    sync --pipfile should work for either dash or underscore names.
+
+    Asserts fix for https://github.com/Madoshakalaka/pipenv-setup/issues/72.
+    """
+    pipfile_dir = shared_datadir / source_pipfile_dirname
+    for filename in ("Pipfile", "Pipfile.lock", "setup.py"):
+        copy_file(pipfile_dir / filename, tmp_path)
+
+    with data(str(pipfile_dir), tmp_path):
```
Comment (bryant-finney) [inline tests/sync_pipfile_test.py:75]: After the above updates:

```suggestion
    with data("dash_or_underscore_0", shared_datadir / "dash_or_underscore_0"):
```

File: tests/data/dash_or_underscore_0/Pipfile [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+name = "pypi"
+url = "https://pypi.org/simple"
+verify_ssl = true
+
+[dev-packages]
+gitdir = "*"
+generic-package = { editable = true, path = "." }
+
+[packages]
+xml_subsetter = "*"
```
Comment (bryant-finney) [inline tests/data/dash_or_underscore_0/Pipfile:11]: To cover the hyphen case:

```suggestion
xml_subsetter = "*"
pytest-cov = "*"
```

**Review by bryant-finney — CHANGES_REQUESTED** [review]
Awesome job! I had a few suggestion; to summarize:

- [ ] `test_sync_underscore_or_dash()` can be simplified a bit further:
  ```py
  def test_sync_underscore_or_dash(shared_datadir):  # type: (Path) -> None
      """
      sync --pipfile should work for either dash or underscore names.

      Asserts fix for https://github.com/Madoshakalaka/pipenv-setup/issues/72.
      """
      with data("dash_or_underscore_0", shared_datadir / "dash_or_underscore_0"):
          cmd(["", "sync", "--pipfile"])
          cmd(["", "check"])
  ```
- [ ] `tests/data/dash_or_underscore_0/Pipfile.lock` can be removed
- [ ] `tests/data/dash_or_underscore_0/pyproject.toml` can be removed
- [ ] add a hyphenated dependency to `tests/data/dash_or_underscore_0/Pipfile`  for completeness

**Comment by jshwi** [comment]
@bryant-finney awesome, thanks for the review, I'll check out you found

File: tests/sync_pipfile_test.py (hunk 2 above)
Comment (jshwi) [reply tests/sync_pipfile_test.py:62]: Ah, gotcha, it makes sense to me now.

**Comment by jshwi** [comment]
@bryant-finney the pyproject.toml was a side effects of `pipenv lock`, apparently it adds those keys now (or creates a new one if it does not exist).

File: tests/sync_pipfile_test.py (hunk 5 above)
Comment (jshwi) [reply tests/sync_pipfile_test.py:71]: You are right, that can only occur if an error is raised and we are not calling this:
```python
with pytest.raises(SystemExist) as err:
    ....
```
We can simply test that an error is not raised

File: tests/sync_pipfile_test.py (hunk 6 above)
Comment (jshwi) [reply tests/sync_pipfile_test.py:66]: awesome, I will dig in a little deeper when it comes to the fixtures in this suite and how the tests are generally structured.

File: tests/data/dash_or_underscore_0/Pipfile (hunk 7 above)
Comment (jshwi) [reply tests/data/dash_or_underscore_0/Pipfile:11]: I was considering this, though I didn't see a side effect of checking for underscores at the same time as checking for dashes. I suppose the assumption was a test for dashes would have already been implemented in another test, and if none of the other tests had failed as a result, then this would have been OK.

**Comment by jshwi** [comment]
I will pull this PR as it is now, and remedy these failed tests.

**Comment by jshwi** [comment]
I left the other commits in here for completeness. Of course, this can be rebased onto the master branch, or merged, etc. Whatever the preferred strategy is.

**Comment by jshwi** [comment]
@bryant-finney I've run the test without the change to `pipenv_setup/inconsistency_checker.py` and it fails, so the test does what we want it to

File: tests/data/dash_or_underscore_0/Pipfile (hunk 7 above)
Comment (bryant-finney) [reply tests/data/dash_or_underscore_0/Pipfile:11]: Yup that's completely valid. I imagine `pip` won't change its standard in this behavior at any time in the near future, so feel free to leave it either way 🙂

**Comment by bryant-finney** [comment]
Great 👍 I'll go ahead with squash+merge!
